README
pybank
=======

A python library that implements the primitive Core Banking system (CBS) emulator - .

Usage:
 from pybank.cbs import CBS
 cbs = CBS(host=ip, port=port)
 cbs.run()

examples/cbs_example.py is an example of core banking system emulator application, that interacts connects to authorization switch through ISO8583:1993 protocol. Accounts, transactions and balances are stored in a local sqlite3 database.


CHANGELOG

Septembet 2017 
	0.36 Cardholder name inquiry (processing code 88) transaction support
	0.35 processing statement request transactions
	0.34 storing transaction history in TRANSACTIONS table

July 2017
	0.33 checking card existense only for financial transactions
	0.32 checking whether the card exists, if no -> add it to DB

June 2017
	0.31 Copy Amount, cardholder billing -> Amount, settlement
	0.30 amount_cardholder_billing > 0 check in process_trxn_debit_account()

May 2017
	0.29 tox support
	0.28 Using ternary conditional operator
	0.27 README usage case change from balance inquiry to authorization request
	0.26 Added some usage stuff to README
	0.25 README -> README.rst
	0.24 recv() and send() routines
	0.23 checking logon response code
	0.22 send logon request
	0.21 repsonse codes changed to 3-digit ones
	0.20 Dummy reversal processing
	0.19 checking balance precision (hardcoded to 2 digits after decimal point)
	0.18 settle_auth_advice() for 120 messages
	0.17 Copy DE24 from original message
	0.16 process_trxn_balance_inquiry() takes currency from DE51, not DE49
	0.15 CARDS and ACCOUNTS tables
	0.14 proper request/repsonse message copying
	0.13 Fixed float amount parsing
	0.12 Database class methods get_card_balance(), insert_card_record(), update_card_balance(), account_exists()
	0.11 Fixed trace() raise() caused by tracetools changes in v0.23
	0.10 Initial setup - code moved from bpc8583 library examples

tests/tests.py
#!/usr/bin/env python

import unittest
import os

from pybank.cbs import CBS
from pybank.db import Database 

class TestBuildTLVTag(unittest.TestCase):
    def setUp(self):
        self.cbs = CBS(host=None, port=None)

    def test_get_float_amount(self):
       self.assertEqual(self.cbs.build_tlv_tag(29, 'JOHN DOE'), '029008JOHN DOE')

class TestGetFloatAmount(unittest.TestCase):
    def setUp(self):
        self.cbs = CBS(host=None, port=None)

    def test_get_float_amount(self):
       self.assertEqual(self.cbs.get_float_amount(3585, 643), 35.85)

    def test_get_float_amount_large(self):
       self.assertEqual(self.cbs.get_float_amount(999999999999, 643), 9999999999.99)

    def test_get_float_amount_small(self):
       self.assertEqual(self.cbs.get_float_amount(13, 643), 0.13)


class TestGetTransactionTypeMnemonic(unittest.TestCase):
    def setUp(self):
        self.cbs = CBS(host=None, port=None)

    def get_purchase_transaction_type_mnemonic(self):
        self.assertEqual(self.cbs.get_transaction_type_mnemonic('000001'), 'PRC')

    def get_unknown_transaction_type_mnemonic(self):
        self.assertEqual(self.cbs.get_transaction_type_mnemonic('990101'), 'OTH')


class TestGetField62Data(unittest.TestCase):
    def setUp(self):
        self.cbs = CBS(host=None, port=None)

    def test_get_field62_data_single_record(self):
        trxn_data = [('D', 1270012, '010000', 100, '2017-09-20 15:12:40.617'), ]
        self.assertEqual(self.cbs.get_field62_data(trxn_data), '170920WDLD000001270012000000000000')

    def test_get_field62_data_multiple_records(self):
        trxn_data = [   ('D', 270000, '010000', 100, '2017-09-20 15:12:40.617'), 
                        ('D', 100000, '000000', 100, '2017-09-20 15:11:07.637'), 
                        ('D', 20000, '990000', 100, '2017-09-20 15:09:46.639'), ]
        self.assertEqual(self.cbs.get_field62_data(trxn_data), '170920WDLD000000270000000000000000170920PRCD000000100000000000000000170920OTHD000000020000000000000000')        


class TestGetMessageLength(unittest.TestCase):
    def setUp(self):
        self.cbs = CBS(host=None, port=None)

    def test_get_message_length_ascii_empty(self):
        self.assertEqual(self.cbs.get_message_length(b''), b'\x00\x00')
 
    def test_get_message_length_ascii_XX(self):
        self.assertEqual(self.cbs.get_message_length(b'XX'), b'\x00\x02')

    def test_get_message_length_ascii_length_28(self):
        self.assertEqual(self.cbs.get_message_length(b'XXXXXXXXXXXXXXXXXXXXXXXXXXXX'), b'\x00\x28')

    def test_get_message_length_ascii_length_216(self):
        self.assertEqual(self.cbs.get_message_length(b'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'), b'\x02\x16')


class TestGetBalanceString(unittest.TestCase):
    def setUp(self):
        self.cbs = CBS(host=None, port=None)

    def test_get_balance_string_empty(self):
        self.assertEqual(self.cbs.get_balance_string('', ''), '')

    def test_get_balance_string_positive(self):
        self.assertEqual(self.cbs.get_balance_string(1234.56, '643'), '007016C000000123456643')

    def test_get_balance_string_positive_no_decimals(self):
        self.assertEqual(self.cbs.get_balance_string(1234, '643'), '007016C000000123400643')

    def test_get_balance_string_negative(self):
        self.assertEqual(self.cbs.get_balance_string(-1234.59, '826'), '007016D000000123459826')


class TestDatabaseInsertCardRecord(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)

    def tearDown(self):
        os.remove(self.db_name)

    def test_insert_valid_card_record(self):
        self.assertTrue(self.db.insert_card_record('8930011234567890', 826, 100500.00))

    def test_insert_duplicate_card_records(self):
        self.assertTrue(self.db.insert_card_record('8930011234567890', 826, 100500.00))
        self.assertFalse(self.db.insert_card_record('8930011234567890', 826, 100500.00))

    def test_insert_differenct_card_records(self):
        self.assertTrue(self.db.insert_card_record('3333333333333333', 826, 100500.00))
        self.assertTrue(self.db.insert_card_record('4444444444444444', 826, 100500.00))    
   
    def test_insert_valid_card_record_check_precision(self):
        self.assertTrue(self.db.insert_card_record('8930011234567890', 826, 100500.00000007))
        self.assertEqual(self.db.get_card_balance('8930011234567890', 826), 100500.0)


class TestDatabaseGetCardBalance(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)
        self.card = '8930011234567890'
        self.currency = 826
        self.db.insert_card_record(self.card, self.currency, 100500.00)

    def tearDown(self):
        os.remove(self.db_name)

    def test_get_card_balance_valid_card(self):
        self.assertEqual(self.db.get_card_balance(self.card, self.currency), 100500.0)

    def test_get_card_balance_currecy_as_string(self):
        self.assertEqual(self.db.get_card_balance(self.card, '826'), 100500.0)

    def test_get_card_balance_wrong_currency(self):
        self.assertEqual(self.db.get_card_balance(self.card, 840), None)

    def test_get_card_balance_inexistent_card(self):
        self.assertEqual(self.db.get_card_balance('801803291823', self.currency), None)

    def test_get_card_balance_card_no_empty(self):
        self.assertEqual(self.db.get_card_balance('', self.currency), None)


class TestDatabaseGetCardDefaultCurrency(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)
        self.card = '8930011234567890'
        self.currency = 826
        self.db.insert_card_record(self.card, self.currency, 100500.00)

    def tearDown(self):
        os.remove(self.db_name)

    def test_get_card_currency(self):
        self.assertEqual(self.db.get_card_default_currency(self.card), self.currency)


class TestDatabaseCardHasAccount(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)
        self.card = '8930011234567890'
        self.currency = 826
        self.db.insert_card_record(self.card, self.currency, 100500.00)

    def tearDown(self):
        os.remove(self.db_name)

    def test_card_has_account(self):
        self.assertTrue(self.db.card_has_account(self.card, self.currency))

    def test_account_doesnt_exists_wrong_card(self):
        self.assertFalse(self.db.card_has_account('192830918039', self.currency))
    
    def test_account_doesnt_exists_wrong_currency(self):
        self.assertFalse(self.db.card_has_account(self.card, 643))


class TestDatabaseCardExists(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)
        self.card = '8930011234567890'
    
    def tearDown(self):
        os.remove(self.db_name)

    def test_card_does_not_exist(self):
        self.assertFalse(self.db.card_exists(self.card))

    def test_card_exists(self):
        self.db.insert_card_record(self.card, 643, 100500.00)
        self.assertTrue(self.db.card_exists(self.card))


class TestDatabaseUpdateCardBalance(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)
        self.card = '8930011234567890'
        self.currency = 826
        self.db.insert_card_record(self.card, self.currency, 100500.00)

    def tearDown(self):
        os.remove(self.db_name)

    def test_update_card_balance(self):
        new_balance = 12.34
        self.assertTrue(self.db.update_card_balance(self.card, self.currency, new_balance))
        self.assertEqual(self.db.get_card_balance(self.card, self.currency), new_balance)

    def test_update_card_balance_wrong_card(self):
        new_balance = 12.34
        self.assertEqual(self.db.update_card_balance('iddqd', self.currency, new_balance), None)

    def test_update_card_balance_wrong_currency(self):
        new_balance = 12.34
        self.assertEqual(self.db.update_card_balance(self.card, 840, new_balance), None)
        self.assertEqual(self.db.get_card_balance(self.card, 840), None)

    def test_update_card_balance_check_precision(self):
        new_balance = 17.890000000003
        self.assertTrue(self.db.update_card_balance(self.card, 826, new_balance))
        self.assertEqual(self.db.get_card_balance(self.card, 826), 17.89)


class TestDatabaseInsertTransactionRecord(unittest.TestCase):
    def setUp(self):
        self.db_name = 'tests.db'
        self.db = Database(self.db_name)
        self.db.insert_card_record('8930011234567890', 826, 100500.00)

    def tearDown(self):
        os.remove(self.db_name)

    def test_insert_valid_transaction_record(self):
        self.db.insert_transaction_record('1110', '8930011234567890', 826, 30.0, 'D', prcode=None);
        
        last_trxns = self.db.get_last_transactions('8930011234567890')
        last_trxn = last_trxns[0]
        self.assertEqual(last_trxn[0], 'D' )
        self.assertEqual(last_trxn[1], 30.0 )
    
    def test_insert_valid_transaction_record_with_prcode(self):
        self.db.insert_transaction_record('1421', '8930011234567890', 826, 23.12, 'D', '010000');
        
        last_trxns = self.db.get_last_transactions('8930011234567890')
        last_trxn = last_trxns[0]
        self.assertEqual(last_trxn[0], 'D' )
        self.assertEqual(last_trxn[1], 23.12 )
        self.assertEqual(last_trxn[2], '010000' )
    
    def test_get_three_last_transactions(self):
        self.db.insert_transaction_record('1110', '8930011234567890', 826, 23.12, 'D', '010000');
        self.db.insert_transaction_record('1420', '8930011234567890', 826, 223.44, 'D');
        self.db.insert_transaction_record('1110', '8930011234567890', 826, 19.04, 'C', '000000');
        
        last_trxns = self.db.get_last_transactions('8930011234567890', 3)
        first = last_trxns[0]
        second = last_trxns[1]
        third = last_trxns[2]

        self.assertEqual(first[0], 'C' )
        self.assertEqual(first[1], 19.04 )
        self.assertEqual(first[2], '000000' )

        self.assertEqual(second[0], 'D' )
        self.assertEqual(second[1], 223.44 )
        self.assertEqual(second[2], None )

        self.assertEqual(third[0], 'D' )
        self.assertEqual(third[1], 23.12 )
        self.assertEqual(third[2], '010000' )

"""
"""

if __name__ == '__main__':
    unittest.main()

pybank/db.py
import sqlite3
import random
from math import pow


class Database:
	def __init__(self, db_name=None):
		self.db_name = db_name if db_name else 'cbs.db'

		self.conn = sqlite3.connect(self.db_name)
		self.cursor = self.conn.cursor()
		try:
			self.conn.execute('select 1 from CARDS')
		except sqlite3.OperationalError:
			self.db_init()


	def db_init(self):
		self.conn = sqlite3.connect(self.db_name)
		self.cursor = self.conn.cursor()
		self.conn.execute('create table CARDHOLDERS (id INTEGER PRIMARY KEY autoincrement, first_name TEXT, last_name TEXT, city TEXT, address TEXT)');
		self.conn.execute('create table ACCOUNTS (account_number INTEGER, type INTEGER default 0, currency INTEGER, balance real, cardholder integer, FOREIGN KEY(cardholder) REFERENCES CARDHOLDERS(id), constraint ACCOUNT_CURRENCY_UK unique(account_number, currency));')
		self.conn.execute('create table CARDS (card_number INTEGER, currency INTEGER, account INTEGER, constraint CARD_CURRENCY_UK unique(card_number, currency), FOREIGN KEY(account) REFERENCES ACCOUNTS(account_number));')
		self.conn.execute("create table TRANSACTIONS (mti number, card_number INTEGER, currency INTEGER, amount INTEGER, debit_credit_flag CHAR not null default 'D', timestamp DATETIME DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')), prcode TEXT, stan TEXT, rrn TEXT, field48 TEXT, FOREIGN KEY(card_number, currency) REFERENCES cards(card_number, currency));")


	def get_card_balance(self, card, currency_code):
		"""
		"""
		t = (card,currency_code)
		self.cursor.execute('select balance from ACCOUNTS where account_number=(select account from CARDS where card_number=? and currency=?)', t)
		row = self.cursor.fetchone()
		return row[0] if row else None


	def get_card_default_currency(self, card):
		t = (card,)
		self.cursor.execute('select currency from CARDS where card_number=?', t)
		row = self.cursor.fetchone()
		return row[0] if row else None


	def generate_account_number(self):
		"""
		"""
		return '999' + str(random.randint(pow(2, 8*2), pow(2, 8*4)))


	def create_account(self, currency, balance):
		"""
		"""
		account_number = self.generate_account_number()
		# TODO: check currency exponent
		t = (account_number, currency, format(balance, '.2f'))
		try:
			self.conn.execute('insert into ACCOUNTS(account_number, currency, balance) values(?,?,?)', t)
			self.conn.commit()
		except (sqlite3.IntegrityError,sqlite3.ProgrammingError) as e:
			print(e)
			return False
		return account_number


	def insert_card_record(self, card, currency, balance):
		"""
		"""
		account_number = self.create_account(currency, balance)
		if not account_number:
			return False

		t = (card, currency, account_number)
		try:
			self.conn.execute('insert into CARDS(card_number, currency, account) values(?,?,?)', t)
			self.conn.commit()
		except (sqlite3.IntegrityError,sqlite3.ProgrammingError) as e:
			print(e)
			return False
		return True


	def update_card_balance(self, card, currency, new_balance):
		"""
		"""
		if self.card_has_account(card, currency):
			try:
				# TODO: check currency exponent
				t = (format(new_balance, '.2f'), card, currency)
				self.cursor.execute('update ACCOUNTS set balance=? where account_number=(select account from cards where card_number=? and currency=?)', t)
				self.conn.commit()
				return True
			except:
				return False
		else:
			return None


	def card_has_account(self, card, currency):
		"""
		"""
		t = (card,currency)
		self.cursor.execute('select 1 from CARDS where card_number=? and currency=?', t)
		row = self.cursor.fetchone()
		return True if row else False


	def card_exists(self, card):
		"""
		"""
		t = (card,)
		self.cursor.execute('select 1 from CARDS where card_number=?', t)
		row = self.cursor.fetchone()
		return True if row else False		


	def insert_transaction_record(self, mti, card, currency, amount, flag, prcode=None, STAN=None, RRN=None, Field48=None):
		"""
		"""
		t = (mti, card, currency, amount, flag, prcode, STAN, RRN, Field48)
		try:
			self.conn.execute('insert into TRANSACTIONS(mti, card_number, currency, amount, debit_credit_flag, prcode, STAN, RRN, Field48) values(?,?,?,?,?,?,?,?,?)', t)
			self.conn.commit()
		except (sqlite3.IntegrityError,sqlite3.ProgrammingError) as e:
			print(e)
			return False
		return True


	def get_last_transactions(self, card, size=1):
		"""
		"""
		t = (card,)
		self.cursor.execute('select debit_credit_flag, amount, prcode, mti, timestamp from TRANSACTIONS where card_number=? order by timestamp desc', t)
		rows = self.cursor.fetchmany(size)
		return rows

	def get_cardholder_name(self, card):
		t = (card,)
		self.cursor.execute('select first_name, last_name from cardholders where id = (select id from cards c inner join accounts a on c.account=a.account_number where c.card_number=?)', t)
		row = self.cursor.fetchone()

		if row:
			return row[0] + ' ' + row[1]


examples/cbs_example.py
#!/usr/bin/env python

import getopt 
import sys

from pybank.cbs import CBS

def show_help(name):
    """
    Show help and basic usage
    """
    print('Usage: python3 {} [OPTIONS]... '.format(name))
    print('ISO8583 Core banking system simulator')
    print('  -p, --port=[PORT]\t\tTCP port to connect to, 3388 by default')
    print('  -s, --server=[IP]\t\tIP of the host to connect to, 127.0.0.1 by default')


if __name__ == '__main__':
    ip = ''
    port = 3388

    optlist, args = getopt.getopt(sys.argv[1:], 'hp:s:', ['help', 'port=', 'server='])
    for opt, arg in optlist:
        if opt in ('-h', '--help'):
            show_help(sys.argv[0])
            sys.exit()
        elif opt in ('-p', '--port'):
            try:
                port = int(arg)
            except ValueError:
                print('Invalid TCP port: {}'.format(arg))
                sys.exit()

        elif opt in ('-s', '--server'):
            ip = arg

    cbs = CBS(host=ip, port=port)
    cbs.run()

pybank/cbs.py
#!/usr/bin/env python

import socket
import sys
import struct
import sqlite3

from binascii import hexlify
from time import sleep

from bpc8583.ISO8583 import ISO8583, MemDump, ParseError
from bpc8583.spec import IsoSpec, IsoSpec1987BPC
from bpc8583.tools import get_response, get_stan, get_datetime_with_year

from tracetools.tracetools import trace
from pynblock.tools import B2raw

from pybank.db import Database

class CBS:
    def __init__(self, host=None, port=None):
        self.host = host if host else '127.0.0.1'

        if port:
            try:
                self.port = int(port)
            except ValueError:
                print('Invalid TCP port: {}'.format(arg))
                sys.exit()
        else:
            self.port = 3388

        self.db = Database('cbs.db')
        self.responses = {'Approval': '000', 'Invalid Amount': '903', 'Invalid account number': '914', 'Insufficient funds': '915', }


    def get_message_length(self, message):
        return B2raw(bytes(str(len(message)).zfill(4), 'utf-8'))


    def build_tlv_tag(self, tag, value):
        return str(tag).zfill(3) + str(len(value)).zfill(3) + str(value)


    def get_balance_string(self, balance, currency_code):
        """
        Get balance string, according to Field 54 description
        """
        if not balance or not currency_code:
            return ''
    
        amount_sign = 'C' if balance > 0 else 'D'
    
        balance_formatted = '{0:.2f}'.format(balance).replace(' ', '').replace('.', '').replace('-', '').zfill(12)
        balance_string = amount_sign + balance_formatted + currency_code
    
        return '007' + str(len(balance_string)).zfill(3) + balance_string


    def get_float_amount(self, amount, currency):
        """
        TODO: check currency exponent, currently using 2 by default
        """
        return amount / 100.0 if amount else .0


    def connect(self):
        """
        """
        try:
            self.sock = None
            for res in socket.getaddrinfo(self.host, self.port, socket.AF_UNSPEC, socket.SOCK_STREAM):
                af, socktype, proto, canonname, sa = res
                self.sock = socket.socket(af, socktype, proto)
                self.sock.connect(sa)
        except OSError as msg:
            print('Error connecting to {}:{} - {}'.format(self.host, self.port, msg))
            sys.exit()
        print('Connected to {}:{}'.format(self.host, self.port))


    def process_trxn_balance_inquiry(self, request, response):
        card_number = request.FieldData(2)
        currency_code = request.FieldData(51)
        if not currency_code:
            currency_code = request.FieldData(49)

        available_balance = self.db.get_card_balance(card_number, currency_code)

        response.FieldData(54, self.get_balance_string(available_balance, currency_code))
        response.FieldData(39, self.responses['Approval'])


    def get_transaction_type_mnemonic(self, prcode):
        trxn_type = prcode[0:2]

        if trxn_type == '00':
            return 'PRC'
        elif trxn_type == '01':
            return 'WDL'
        else:
            return 'OTH'


    def get_field62_data(self, trxn_data):
        """
        """
        field_data = ''
        print_data = '\tStatement data:\n'

        for record in trxn_data:
            # Date
            # '2017-09-20 15:12:40.617' -> '170920'
            date = record[4].split(' ')[0].split('-')
            field_data += date[0][2:4] + date[1] + date[2]
            print_data += record[4] + '\t'

            # Transaction type
            field_data += self.get_transaction_type_mnemonic(record[2])
            print_data += self.get_transaction_type_mnemonic(record[2]) + '\t'

            # Debit/Credit flag
            field_data += record[0]
            print_data += record[0] + ' '

            # Amount
            amount = str(record[1]).zfill(12)
            field_data += amount
            print_data += amount + '\n'

            # Dummy Fee
            fee_amount = '000000000000'
            field_data += fee_amount

        print(print_data)
        return field_data

    def process_cardholder_name_inquiry(self, request, response):
        """
        """
        card_number = request.FieldData(2)
        cardholder_name = self.db.get_cardholder_name(card_number)
        response.FieldData(48, self.build_tlv_tag(29, cardholder_name))
        response.FieldData(39, self.responses['Approval'])


    def process_statement_request(self, request, response):
        """
        """
        card_number = request.FieldData(2)

        trxns = self.db.get_last_transactions(card_number, 10)

        response.FieldData(39, self.responses['Approval'])
        response.FieldData(48, None)
        response.FieldData(62, self.get_field62_data(trxns))


    def save_transaction(self, request, flag='D'):
        """
        """
        MTI = str(request.get_MTI()).zfill(4)[-3:]
        card_number = request.FieldData(2)
        billing_currency_code = request.FieldData(51)
        amount_cardholder_billing = self.get_float_amount(request.FieldData(6), billing_currency_code)
        prcode = str(request.FieldData(3)).zfill(6)
        STAN = request.FieldData(11)
        RRN = request.FieldData(37)
        Field48 = request.FieldData(48)

        if amount_cardholder_billing:
            self.db.insert_transaction_record(mti=MTI, card=card_number, currency=billing_currency_code, amount=amount_cardholder_billing, flag=flag, prcode=prcode, STAN=STAN, RRN=RRN, Field48=Field48);


    def process_trxn_debit_account(self, request, response):
        """
        """
        card_number = request.FieldData(2)
        transaction_currency_code = request.FieldData(49)
        billing_currency_code = request.FieldData(51)
        amount_cardholder_billing = self.get_float_amount(request.FieldData(6), billing_currency_code)

        available_balance = self.db.get_card_balance(card_number, billing_currency_code)

        print('Trying to debit customer. Card [{}], currency [{}], balance before[{}]\n'.format(card_number, billing_currency_code, available_balance));

        if available_balance:

            if transaction_currency_code == billing_currency_code and not amount_cardholder_billing:
                amount_cardholder_billing = self.get_float_amount(request.FieldData(4), billing_currency_code)

            if amount_cardholder_billing:
                if available_balance > amount_cardholder_billing:
                    self.db.update_card_balance(card_number, billing_currency_code, available_balance - amount_cardholder_billing)

                    response.FieldData(39, self.responses['Approval'])
    
                    available_balance = self.db.get_card_balance(card_number, billing_currency_code)
                    response.FieldData(54, self.get_balance_string(available_balance, billing_currency_code))
                    print('Debited succesfully');
                    self.save_transaction(request, 'D')

                else:
                    response.FieldData(39, self.responses['Insufficient funds'])
                    response.FieldData(54, self.get_balance_string(available_balance, billing_currency_code))
                    print('Debit failed: insufficient funds');
            else:
                response.FieldData(39, self.responses['Invalid Amount'])
                print('Invalid Cardholder billing amount')
        else:
            response.FieldData(39, self.responses['Invalid account number'])
            print('Invalid account number')


    def settle_auth_advice(self, request, response):
        """
        """
        card_number = request.FieldData(2)
        currency_code = request.FieldData(51)
        amount_cardholder_billing = self.get_float_amount(request.FieldData(6), currency_code)

        available_balance = self.db.get_card_balance(card_number, currency_code)
        self.db.update_card_balance(card_number, currency_code, available_balance - amount_cardholder_billing)
        self.save_transaction(request, 'D')

        response.FieldData(39, self.responses['Approval'])


    def settle_reversal(self, request, response):
        """
        """
        card_number = request.FieldData(2)
        currency_code = request.FieldData(51)
        amount_cardholder_billing = self.get_float_amount(request.FieldData(6), currency_code)

        available_balance = self.db.get_card_balance(card_number, currency_code)
        self.db.update_card_balance(card_number, currency_code, available_balance + amount_cardholder_billing)
        self.save_transaction(request, 'C')
        response.FieldData(39, self.responses['Approval'])


    def init_response_message(self, request):
        """
        """
        response = ISO8583(None, IsoSpec1987BPC())
        response.MTI(get_response(request.get_MTI()))

        # Copy some key fields from original message:
        for field in [2, 3, 4, 5, 6, 11, 12, 14, 15, 17, 24, 32, 37, 48, 49, 50, 51, 102]:
            response.FieldData(field, request.FieldData(field))

        # Copy Amount, cardholder billing -> Amount, settlement
        response.FieldData(5, request.FieldData(6))

        # Copy Currency code, cardholder billing -> Currency code, settlement
        response.FieldData(50, request.FieldData(51))

        return response


    def get_transaction_type(self, request):
        """
        """
        if request.FieldData(3) != None:
            processing_code = str(request.FieldData(3)).zfill(6)
            return processing_code[0:2]
        else:
            return None


    def send_logon_handshake(self):
        """
        """
        while True:
            request = ISO8583(None, IsoSpec1987BPC())
    
            request.MTI('1804')
            request.FieldData(11, get_stan())
            request.FieldData(12, get_datetime_with_year())
            request.FieldData(24, 801)
    
            request.Print()
            data = request.BuildIso()
            data = self.get_message_length(data) + data
            self.send(data)
    
            data = self.recv()
            response = ISO8583(data, IsoSpec1987BPC())
            response.Print()
    
            if response.get_MTI() != '1814' or response.FieldData(24) != 801 or response.FieldData(39) != '000':
                print('\tLogon failed')
                sleep(5)
            else:
                break


    def recv(self):
        """
        """
        data = self.sock.recv(4096)
        if len(data) > 0:
            trace(title='<< {} bytes received: '.format(len(data)), data=data)
            return data[2:]
        else:
            return None


    def send(self, data):
        """
        """
        self.sock.send(data)
        trace(title='>> {} bytes sent:'.format(len(data)), data=data)


    def run(self):
        """
        """
        while True:
            try:
                self.connect()
                self.send_logon_handshake()

                while True:
                    data = self.recv() 
                    if not data:
                        raise ParseError

                    request = ISO8583(data, IsoSpec1987BPC())
                    request.Print()

                    response = self.init_response_message(request)

                    MTI = str(request.get_MTI()).zfill(4)[-3:]
                    trxn_type = self.get_transaction_type(request)
                    card_number = request.FieldData(2)
                    currency_code = request.FieldData(51)

                    if MTI in ['100', '200']:
                        # Authorization request or financial request
                        if not self.db.card_exists(card_number):
                            # TODO: command line option to add cards autmatically
                            print('\nWARNING: Card {} does not exist!\n'.format(card_number));
                            self.db.insert_card_record(card_number, currency_code, 0);

                        if trxn_type == '31':
                            # Balance
                            self.process_trxn_balance_inquiry(request, response)
                        elif trxn_type in ['00', '01', '50']:
                            # Purchase or ATM Cash
                            self.process_trxn_debit_account(request, response)
                        elif trxn_type == '39':
                            # Ministatement
                            self.process_statement_request(request, response)
                        elif trxn_type == '88':
                            # Cardholder name inquiry
                            self.process_cardholder_name_inquiry(request, response)
                        else:
                            print('Unsupported transaction type {}. Responding APPROVAL by default'.format(trxn_type))
                            response.FieldData(39, self.responses['Approval'])

                    elif MTI in ['120']:
                        # Authorization advice
                        if trxn_type in ['00', '01']:
                            # Purchase or ATM Cash
                            self.settle_auth_advice(request, response)
                        else:
                            response.FieldData(39, self.responses['Approval'])

                    elif MTI in ['400', '420']:
                        # Reversal
                        if trxn_type in ['00', '01']:
                            # Purchase or ATM Cash
                            self.settle_reversal(request, response)

                    response.Print()
                    
                    data = response.BuildIso()
                    data = self.get_message_length(data) + data
                    self.send(data)
        
            except ParseError:
                print('Connection closed')
                conn.close()

        self.sock.close()
        conn.close()
